pallets/migration/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+	use sp_version::RuntimeVersion;
+
+	// Simple declaration of the `Pallet` type. It is placeholder we use to implement traits and
+	// method.
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Upgrader structure that provides the actual logic for the upgrades
+		type Upgrades: RuntimeUpgradeProvider;
+
+		/// Used in order to convert the given weights into a number of blocks
+		type WeightToBlockNumber: Convert<Weight, Self::BlockNumber>;
+
+		/// Associated type for Event enum
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			// TODO: Insert some mechanism to prevent the accidential execution of the same
+			// 	upgreade twice. Probably best, to let the runtime version be the key for the
+			// 	map of old upgrades and if it already exists, do not do the upgrade!
+			let max_per_block = T::BlockWeights::get().max_block;
+
+			// In order to safely unwrap below, we do this check. Although, one might argue
+			// a chain with a maximum weight of zero, might be useless.
+			if max_per_block == <Weight as Zero>::zero() {
+				// As we are not updating the storage, this will basically prevent the upgrade from
+				// running.
+				return Zero::zero();
+			}
+
+			// The number of blocks we are going to block the chain for the upgreade
+			//
+			// NOTE:
+			// For the reasons of missing a remainder we are adding 1 extra block at the end,
+			// so that the upgreade will definitely fit into the given blocks.
+			let period: T::BlockNumber = <T::WeightToBlockNumber>::convert(
+				T::Upgrades::upgrade_weight()
+					.checked_div(max_per_block)
+					.expect("Maximum block weight is not zero. qed."),
+			)
+			.saturating_add(One::one());
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			// Insert start and period into storage for the `on_initialize`
+			<UpgradeStart<T>>::set(now);
+			<UpgradePeriod<T>>::set(period);
+
+			let (version, memo) = T::Upgrades::info();
+			<UpgradeInfo<T>>::insert(now, (version.clone(), Vec::from(memo.as_ref())));
+
+			Self::deposit_event(Event::RuntimeUpgradeStarted(period, version));
+
+			// return zero as `on_initialize` will take care of the upgrade
+			Zero::zero()
+		}
+
+		fn on_initialize(n: T::BlockNumber) -> frame_support::weights::Weight {
+			// You can not do an upgrade on the first block...
+			if n < Self::upgrade_start().saturating_add(Self::upgrade_period()) && n != Zero::zero()
+			{
+				let period = Self::upgrade_period();
+				let start = Self::upgrade_start();
+				let max_per_block = T::BlockWeights::get().max_block;
+
+				Self::upgrade(max_per_block);
+
+				// Indicate to the network, that the upgrade has been executed successfully
+				if (n == start.saturating_add(period) - One::one())
```

Comment:
I think we should not restrict to a specific # of blocks for upgrade but rather batch the data within a set weight and let it run until batches are complete and then emit an event maybe?

Reply:
I think, we need to block whole blocks, even if it is just for the reason that everybody knows, for `n` blocks there won't be any extrinsic coming through. And with this, we can always block some extra space. 

Or do you have a specific reason, to go for the weights directly?